File: tools_box/_stock/report/stock_valuation_report/stock_valuation_report.js
// Copyright (c) 2016, masonarmani38@gmail.com and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Stock Valuation Report"] = {
	"filters": [
{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"default": frappe.datetime.year_start()
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"default": frappe.datetime.year_end()
		},
		{
			"fieldname":"warehouse",
			"label": __("Warehouse"),
			"fieldtype": "Link",
			"options": "Warehouse"
		}
	]
}

File: tools_box/_stock/report/stock_valuation_report/stock_valuation_report.py
# Copyright (c) 2015, masonarmani38@gmail.com. and Contributors
# License: GNU General Public License v3. See license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import flt, cint, getdate


def execute(filters=None):
    if not filters: filters = {}

    columns = get_columns()
    data = []

    conditions = "posting_date BETWEEN DATE('{0}') AND DATE('{1}')".format(filters.get('from_date'),
                                                                           filters.get('to_date'))

    if filters.get('warehouse'):
        conditions += " AND warehouse ='{0}'".format(filters.get('warehouse'))

    # get all active items and get last valuation rate
    items = frappe.db.sql("SELECT name, item_name, item_group, brand, description , stock_uom FROM `tabItem` "
                          "WHERE disabled = 0", as_dict=1)
    for item in items:
        sle = frappe.db.sql(
            "SELECT name, valuation_rate, company, warehouse FROM `tabStock Ledger Entry` "
            "WHERE item_code = '{0}' AND {1} ORDER BY posting_date DESC  LIMIT 1 ".format(item.name, conditions),
            as_dict=1)

    if sle:
        mp = (
            item.name, item.item_name, item.item_group, item.description, sle[0].get('warehouse'),
            item.stock_uom, sle[0].get('valuation_rate'), sle[0].get('name'), sle[0].get('company'),
        )
    data.append(mp)

    return columns, data


def get_columns():
    """return columns"""
    columns = [
        _("Item") + ":Link/Item:100",
        _("Item Name") + "::150",
        _("Item Group") + "::100",
        _("Description") + "::170",
        _("Warehouse") + ":Link/Warehouse:100",
        _("Stock UOM") + ":Link/UOM:90",
        _("Valuation Rate") + ":Float:90",
        _("Last Stock Ledger Entry") + ":Link/Stock Ledger Entry:120",
        _("Company") + ":Link/Company:100"
    ]

    return columns
